pkg/ferry-controller/controller/hub/cluster_serviceexports_cache.go

```diff
@@ -0,0 +1,184 @@
+/*
+Copyright 2022 FerryProxy Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hub
+
+import (
+	"context"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/ferryproxy/ferry/pkg/utils/objref"
+	"github.com/ferryproxy/ferry/pkg/utils/trybuffer"
+	"github.com/go-logr/logr"
+	"k8s.io/client-go/tools/cache"
+	"sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
+	"sigs.k8s.io/mcs-api/pkg/client/clientset/versioned"
+	"sigs.k8s.io/mcs-api/pkg/client/informers/externalversions"
+)
+
+type clusterServiceExportCache struct {
+	parentCtx context.Context
+	ctx       context.Context
+	cancel    context.CancelFunc
+
+	clientset versioned.Interface
+	cache     map[objref.ObjectRef]*v1alpha1.ServiceExport
+	syncFunc  func()
+
+	logger logr.Logger
+	try    *trybuffer.TryBuffer
+
+	mut sync.RWMutex
+}
+
+type clusterServiceExportCacheConfig struct {
+	Clientset versioned.Interface
+	Logger    logr.Logger
+	SyncFunc  func()
+}
+
+func newClusterServiceExportCache(conf clusterServiceExportCacheConfig) *clusterServiceExportCache {
+	c := &clusterServiceExportCache{
+		clientset: conf.Clientset,
+		logger:    conf.Logger,
+		syncFunc:  conf.SyncFunc,
+		cache:     map[objref.ObjectRef]*v1alpha1.ServiceExport{},
+	}
+	return c
+}
+
+func (c *clusterServiceExportCache) ResetClientset(clientset versioned.Interface) error {
+	c.mut.Lock()
+	defer c.mut.Unlock()
+
+	c.cache = map[objref.ObjectRef]*v1alpha1.ServiceExport{}
+	if c.cancel != nil {
+		c.cancel()
+	}
+	c.ctx, c.cancel = context.WithCancel(c.parentCtx)
+	c.clientset = clientset
+	informerFactory := externalversions.NewSharedInformerFactoryWithOptions(c.clientset, 0)
+	informer := informerFactory.
+		Multicluster().
+		V1alpha1().
+		ServiceExports().
+		Informer()
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.onAdd,
+		UpdateFunc: c.onUpdate,
+		DeleteFunc: c.onDelete,
+	})
+
+	go informer.Run(c.ctx.Done())
+	return nil
+}
```

Comment:
@kerthcet 
When the cluster's Kubeconfig is updated, the Clientset will need to be rebuilt using the parameters associated with its creation, needs `c.parentCtx` and `clientset versioned.Interface`